### PR TITLE
Add New Methods to help with system orchestration

### PR DESCRIPTION
**Pull request by PineappleIOnic**
+ Added getCPUCores()
+ Added UNIX Helper functions
+ Added getCPUUtilisation
+ Added getMemoryTotal()
+ Added getMemoryFree()
+ Added getDiskTotal()
+ Added getDiskFree()
+ Added getDiskStats()
+ Added getIOUsage()
+ Added getNetworkUsage()

Most of these methods are currently Linux only and are currently being iterated on to improve stability.

### CODE REVIEW

File: src/System/System.php [hunk 1]
```diff
@@ -128,4 +155,351 @@ static public function isArch(string $arch): bool
...
+    {
+        // Read /proc/diskstats
+        $diskstats = file_get_contents('/proc/diskstats');
+
+        // Split the data
+        $diskstats = explode("\n", $diskstats);
+
+        // Remove excess spaces
+        $diskstats = array_map(function ($data) {
+            return preg_replace('/\s+/', ' ', trim($data));
```
Comment (wess) [inline src/System/System.php:392]: no need for \t?

**Comment by eldadfux** [comment]
@PineappleIOnic lets add a lot of tests :)

File: src/System/System.php [hunk 2]
```diff
@@ -128,4 +157,358 @@ static public function isArch(string $arch): bool
...
+            return true;
+        });
+
+        // Get the total IO Usage
+        $IOUsage = [];
+
+        foreach ($interfaces as $interface) {
+            $tx1 = intval(file_get_contents('/sys/class/net/' . $interface . '/statistics/tx_bytes'));
+            $rx1 = intval(file_get_contents('/sys/class/net/' . $interface . '/statistics/rx_bytes'));
+            sleep(1);
```
Comment (Meldiron) [inline src/System/System.php:501]: Same as comment above, can we make this attribute?

File: tests/System/SystemTest.php [hunk 3]
```diff
@@ -41,4 +41,69 @@ public function testOs()
         $this->expectException("Exception");
         System::isArch("throw");
     }
+
+    public function testGetCPUCores()
+    {
+        $this->assertIsInt(System::getCPUCores());
```
[the remark this answers is not in the document]
Comment (PineappleIOnic) [reply tests/System/SystemTest.php:47]: Unfortunately, this isn't possible as docker does not write the container's information into the kernel files we read but instead writes the host's usage.

File: src/System/System.php [hunk 4]
```diff
@@ -128,4 +157,360 @@ static public function isArch(string $arch): bool
...
+            $data[$cpuNumber]['nice'] = $cpu[2];
+            $data[$cpuNumber]['system'] = $cpu[3];
+            $data[$cpuNumber]['idle'] = $cpu[4];
+            $data[$cpuNumber]['iowait'] = $cpu[5];
+            $data[$cpuNumber]['irq'] = $cpu[6];
+            $data[$cpuNumber]['softirq'] = $cpu[7];
+
+            // These might not exist on older kernels.
+            $data[$cpuNumber]['steal'] = isset($cpu[8]) ? $cpu[8] : 0;
+            $data[$cpuNumber]['guest'] = isset($cpu[9]) ? $cpu[9] : 0;
```
Comment (eldadfux) [inline src/System/System.php:225]: We can use the null safety check and give proper default where needed.

File: src/System/System.php [hunk 5]
```diff
@@ -128,4 +157,360 @@ static public function isArch(string $arch): bool
...
+                    $i++;
+                }
+
+                if ($id === 0) {
+                    return intval(array_sum($data));
+                } else {
+                    return $data[$id];
+                }
+            default:
+                throw new Exception(self::getOS() . " not supported.");
```
Comment (eldadfux) [inline src/System/System.php:302]: Would be nice to add a table covering all the different method that are not supported in all operating systems.

File: src/System/System.php [hunk 6]
```diff
@@ -128,4 +157,360 @@ static public function isArch(string $arch): bool
...
+     * 
+     * @throws Exception
+     */
+    static public function getMemoryFree(): int
+    {
+        switch (self::getOS()) {
+            case 'Linux':
+                $meminfo = file_get_contents('/proc/meminfo');
+                preg_match('/MemFree:\s+(\d+)/', $meminfo, $matches);
+                return intval($matches[1] / 1024);
```
Comment (eldadfux) [inline src/System/System.php:342]: Accessing the array key directly without validating it exists is risky. Let's have a null safety check.

File: src/System/System.php [hunk 7]
```diff
@@ -128,4 +157,360 @@ static public function isArch(string $arch): bool
...
+    static public function getIOUsage($duration = 1): array
+    {
+        $diskStat = self::getDiskStats();
+        sleep($duration);
+        $diskStat2 = self::getDiskStats();
+
+        // Remove invalid disks
+        $diskStat = array_filter($diskStat, function ($disk) {
+            foreach (self::InvalidDisks as $filter) {
+                if (str_contains($disk[2], $filter)) {
```
Comment (eldadfux) [inline src/System/System.php:439]: Same about null safety. Lets make sure we take care of it across the library.

**Review by eldadfux — CHANGES_REQUESTED** [review]

**Comment by eldadfux** [comment]
Tests are failing.

**Comment by PineappleIOnic** [comment]
The tests failing appear to be a Travis issue as they complain about an argon2 dependency; all other tests are passing

**Review by eldadfux — APPROVED** [review]

**Pull request merged**